### PR TITLE
feat(mme): ext-ue-ambr IE field addition.

### LINKED ISSUE #4072: Magma NSA support
This issue tracks the activity related to:
**NSA support with MME + OAI CUPS SPGW + OAI HSS**

Add support for E-RAB Modification Response and Dual Connectivity with New Radio IEs

**NSA support in MME + Magma Subscriberdb + Embedded SPGW**
Ref: http://www.eventhelix.com/5G/non-standalone-access-en-dc/en-dc-secondary-node-addition-core-network-details.pdf

- Update Location Request/Answer: Add "NR as Secondary RAT", "Extended-Max-Requested-BW-UL" and "Extended-Max-Requested-BW-DL"
- Create Session Request/Response: Add "Extended APN-AMBR"
- Activate Default Bearer Request: Add "Extended UE-AMBR Downlink" and "Extended UE-AMBR Uplink"

### PR DESCRIPTION
**Pull request by Vyas20**
Addition of IE extension in UE-AMBR over S1AP message "Activate default EPS bearer context request".

For values higher than 10 gbps, according to 3GPP document 36.413 v15.3.0 (IE 9.2.1.20) there should be an extension field for UE-AMBR. Upon filling this extension filed, the normal value (UE-AMBR) shall be ignored.

This change is tested with S1AP-Tester and works with UE-AMBR values lesser than 4294967295 bps (highest possible value for 32 bit integer). After commenting out the code at "/lte/gateway/python/magma/subscriberdb/rpc_servicer.py" , this feature works for values higher than 4294967295 bps also.

Wireshark  logs are attached here-in.
[ext-ue-ambr.zip](https://github.com/magma/magma/files/8108673/ext-ue-ambr.zip)

### CODE REVIEW

**Comment by uri200** [comment]
Hey @Vyas20, those changes may potentially break some things in FEG.

Can you run `docker-compose -c` at `/feg/gateway/docker` and make sure test passes

Can you also add a test in FEG for those values 
http://github.com/magma/magma/blob/929f524eb2c2aed3ed544091fde224d7f727f215/feg/gateway/services/s6a_proxy/servicers/s6a_proxy_test.go#L129-L130

**Comment by Vyas20** [comment]
> Hey @Vyas20, those changes may potentially break some things in FEG.
> 
> Can you run `docker-compose -c` at `/feg/gateway/docker` and make sure test passes
> 
> Can you also add a test in FEG for those values [http://github.com/magma/magma/blob/929f524eb2c2aed3ed544091fde224d7f727f215/feg/gateway/services/s6a_proxy/servicers/s6a_proxy_test.go#L129-L130](https://github.com/magma/magma/blob/929f524eb2c2aed3ed544091fde224d7f727f215/feg/gateway/services/s6a_proxy/servicers/s6a_proxy_test.go#L129-L130)

I ran  docker-compose build and it worked fine but docker-compose up is getting stuck in a loop for magmad service

"magmad           | ERROR:root:Metrics upload error for service magmad (chunk 0)! [StatusCode.UNAVAILABLE] Received http2 header with status: 502
magmad           | ERROR:root:Metrics upload error for service session_proxy (chunk 0)! [StatusCode.UNAVAILABLE] Received http2 header with status: 502
"

File: feg/protos/s6a_proxy.proto [hunk 1]
```diff
@@ -217,9 +217,9 @@ message UpdateLocationAnswer {
 
     message AggregatedMaximumBitrate {
         // Maximum uplink bitrate
-        uint32 max_bandwidth_ul = 1;
+        uint64 max_bandwidth_ul = 1;
         // Maximum downlink bitrate
-        uint32 max_bandwidth_dl = 2;
+        uint64 max_bandwidth_dl = 2;
```
Comment (ulaskozat) [inline feg/protos/s6a_proxy.proto]: did not we add the unit information to cover cases for >4gbps? what is the reason for this change then?

File: lte/gateway/c/core/oai/test/s1ap_task/test_s1ap_mme_handlers.cpp [hunk 2]
```diff
@@ -578,12 +579,56 @@ TEST_F(S1apMmeHandlersTest, HandleConnectionEstCnf) {
...
+      0x00, 0x04, 0x02, 0x01, 0xd0, 0x11, 0x40, 0x08, 0x04, 0x02, 0x60,
+      0x04, 0x00, 0x02, 0x1c, 0x00, 0x00, 0x43, 0x00, 0x06, 0x00, 0x00,
+      0xf1, 0x10, 0x00, 0x01, 0x00, 0x64, 0x40, 0x08, 0x00, 0x00, 0xf1,
+      0x10, 0x00, 0x00, 0x00, 0xa0, 0x00, 0x86, 0x40, 0x01, 0x30};
+
+  ASSERT_EQ(simulate_pdu_s1_message(initial_ue_bytes, sizeof(initial_ue_bytes),
+                                    state, assoc_id, stream_id),
+            RETURNok);
+
+handle_mme_ue_id_notification(state, assoc_id);
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/s1ap_task/test_s1ap_mme_handlers.cpp]: did you do format? indentations look off to me.

**Comment by ulaskozat** [comment]
proper way of doing this is to actually pick the units and making conversions internally. Do not get the point of delivering this for s1ap integ test only and not be able to use with an actual cloud config.

**Review by ulaskozat — CHANGES_REQUESTED** [review]
see my comments.

File: feg/protos/s6a_proxy.proto [hunk 3]
```diff
@@ -217,9 +217,9 @@ message UpdateLocationAnswer {
 
     message AggregatedMaximumBitrate {
         // Maximum uplink bitrate
-        uint32 max_bandwidth_ul = 1;
+        uint64 max_bandwidth_ul = 1;
```
Comment (ardzoht) [inline feg/protos/s6a_proxy.proto]: we also have the same message definition on `apn.proto` file, which are used on sessiond and subscriberdb message requests. we should update those as well or look into consolidating both into one, @uri200 thoughts?

**Comment by ardzoht** [comment]
@Vyas20 `After commenting out the code at "/lte/gateway/python/magma/subscriberdb/rpc_servicer.py"` what does this part mean? Is this part of the change ? We should not be commenting out code for the purpose of testing rather updating it or removing if not needed.

File: feg/protos/s6a_proxy.proto (hunk 3 above)
Comment (uri200) [reply feg/protos/s6a_proxy.proto]: I agree. That should be changed too. 
It would be good to modify a lte integ test to use those values and see if something breaks.

File: feg/protos/s6a_proxy.proto (hunk 1 above)
Comment (Vyas20) [reply feg/protos/s6a_proxy.proto]: I changed the datatypes because I am using etc/magma/subscriberdb.yml to enter default ambr rates. These changes need not to be a part of this PR and I will revert.

**Comment by Vyas20** [comment]
> @Vyas20 `After commenting out the code at "/lte/gateway/python/magma/subscriberdb/rpc_servicer.py"` what does this part mean? Is this part of the change ? We should not be commenting out code for the purpose of testing rather updating it or removing if not needed.

I have commented out lines 116 onwardds to test the PR and have attached the diff here-in.

[ExtUE-AMBR.txt](https://github.com/magma/magma/files/8156867/ExtUE-AMBR.txt)

File: lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c [hunk 4]
```diff
@@ -915,7 +918,45 @@ void s1ap_handle_conn_est_cnf(
   asn_uint642INTEGER(
       &ie->value.choice.UEAggregateMaximumBitrate.uEaggregateMaximumBitRateUL,
       conn_est_cnf_pP->ue_ambr.br_ul);
-  ASN_SEQUENCE_ADD(&out->protocolIEs.list, ie);
+
+
+if (conn_est_cnf_pP->ue_ambr.br_dl > 10000000000
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c]: can we use a constant or a macro to define these limit values?

File: lte/gateway/c/core/oai/test/s1ap_task/s1ap_mme_test_utils.cpp [hunk 5]
```diff
@@ -126,9 +127,14 @@ status_code_e send_conn_establishment_cnf(mme_ue_s1ap_id_t ue_id,
   establishment_cnf_p->transport_layer_address[0] = bfromcstr("test");
   establishment_cnf_p->gtp_teid[0] = 1;
 
+if(extueambr){
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/s1ap_task/s1ap_mme_test_utils.cpp]: please apply clang-format on all of the included files

File: lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c [hunk 6]
```diff
@@ -915,7 +918,45 @@ void s1ap_handle_conn_est_cnf(
...
+
+ if (conn_est_cnf_pP->ue_ambr.br_ul > 10000000000) {
+  ie_ambrext = (S1ap_UEAggregate_MaximumBitrates_ExtIEs_t*) calloc(
+      1, sizeof(S1ap_UEAggregate_MaximumBitrates_ExtIEs_t));
+  ie_ambrext->id          = S1ap_ProtocolIE_ID_id_extended_uEaggregateMaximumBitRateUL;
+  ie_ambrext->criticality = S1ap_Criticality_ignore;
+  ie_ambrext->extensionValue.present = S1ap_UEAggregate_MaximumBitrates_ExtIEs__extensionValue_PR_ExtendedBitRate ;
+  asn_uint642INTEGER(
+      &ie->value.choice.UEAggregateMaximumBitrate.uEaggregateMaximumBitRateUL,
+      1000000000);
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c]: use constants instead

File: feg/gateway/services/s6a_proxy/servicers/s6a_definitions.go [hunk 7]
```diff
@@ -149,10 +149,10 @@ type AIA struct {
 //	Result-Code {Code:268,Flags:0x40,Length:12,VendorId:0,Value:Unsigned32{2001}}
 //
 type AMBR struct {
-	MaxRequestedBandwidthUL uint32 `avp:"Max-Requested-Bandwidth-UL"`
-	MaxRequestedBandwidthDL uint32 `avp:"Max-Requested-Bandwidth-DL"`
-	ExtendMaxRequestedBwUL  uint32 `avp:"Extended-Max-Requested-BW-UL"`
-	ExtendMaxRequestedBwDL  uint32 `avp:"Extended-Max-Requested-BW-DL"`
+	MaxRequestedBandwidthUL uint64 `avp:"Max-Requested-Bandwidth-UL"`
```
Comment (ardzoht) [inline feg/gateway/services/s6a_proxy/servicers/s6a_definitions.go]: as we accorded, let's keep `uint64` updates on a separate PR

**Review by ardzoht — CHANGES_REQUESTED** [review]

File: lte/gateway/c/core/oai/test/s1ap_task/test_s1ap_mme_handlers.cpp (hunk 2 above)
Comment (Vyas20) [reply lte/gateway/c/core/oai/test/s1ap_task/test_s1ap_mme_handlers.cpp]: Resolved

File: lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c (hunk 4 above)
Comment (Vyas20) [reply lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c]: Resolved

File: lte/gateway/c/core/oai/test/s1ap_task/s1ap_mme_test_utils.cpp (hunk 5 above)
Comment (Vyas20) [reply lte/gateway/c/core/oai/test/s1ap_task/s1ap_mme_test_utils.cpp]: Resolved

File: lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c (hunk 6 above)
Comment (Vyas20) [reply lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c]: Resolved

File: lte/gateway/c/core/oai/test/s1ap_task/s1ap_mme_test_utils.cpp [hunk 8]
```diff
@@ -107,6 +107,7 @@ status_code_e send_s1ap_erab_rel_cmd(s1ap_state_t* state,
 }
 
 status_code_e send_conn_establishment_cnf(mme_ue_s1ap_id_t ue_id,
+                                          bool extueambr,
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/s1ap_task/s1ap_mme_test_utils.cpp]:
```suggestion
                                          bool trigger_ext_ue_ambr,
```

File: lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c [hunk 9]
```diff
@@ -915,7 +920,45 @@ void s1ap_handle_conn_est_cnf(
   asn_uint642INTEGER(
       &ie->value.choice.UEAggregateMaximumBitrate.uEaggregateMaximumBitRateUL,
       conn_est_cnf_pP->ue_ambr.br_ul);
+
+  if (conn_est_cnf_pP->ue_ambr.br_dl > EXT_UE_AMBR_DL ||
+      conn_est_cnf_pP->ue_ambr.br_ul > EXT_UE_AMBR_UL) {
+    extension = calloc(1, sizeof(S1ap_ProtocolExtensionContainer_7327P134_t));
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c:925]: I assume the freeing of this IE is already handled while dec/encoding?

**Review by ardzoht — CHANGES_REQUESTED** [review]
Last minor change requested. thanks!

File: lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c [hunk 10]
```diff
@@ -76,6 +76,8 @@
 #include "lte/gateway/c/core/oai/tasks/s1ap/s1ap_common.h"
 #include "orc8r/gateway/c/common/service303/includes/MetricsHelpers.h"
 
+#define EXT_UE_AMBR_UL 1000000000
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c]: Currently these macros used only in one function, can we move these declaration to function?

Comment (ardzoht) [reply lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c]: @rsarwad IMO even if they are used as of now on only one function, using the macros here or constants help to describe the usage of them instead of the numbers

File: lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c (hunk 9 above)
Comment (Vyas20) [reply lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c:925]: Message is freed later in the fucntion by
 bstring b = blk2bstr(buffer_p, length);
free(buffer_p);

File: lte/gateway/c/core/oai/test/s1ap_task/s1ap_mme_test_utils.cpp (hunk 8 above)
Comment (Vyas20) [reply lte/gateway/c/core/oai/test/s1ap_task/s1ap_mme_test_utils.cpp]: Updated.

**Review by ardzoht — APPROVED** [review]
lgtm

File: lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c [hunk 11]
```diff
@@ -76,6 +76,8 @@
...
+#define EXT_UE_AMBR_DL 1000000000
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c]: should not these be 10Gbps? I am counting 9 zeros here.

**Review by ulaskozat — APPROVED** [review]
I stamped but please check my comment for 1G vs 10G.

File: lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c (hunk 11 above)
Comment (Vyas20) [reply lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_nas_procedures.c]: Updated
